src/zos-mkdtemp.c

```diff
@@ -0,0 +1,105 @@
+//////////////////////////////////////////////////////////////////////////////
+//// Licensed Materials - Property of IBM
+//// ZOSLIB
+//// (C) Copyright IBM Corp. 2020. All Rights Reserved.
+//// US Government Users Restricted Rights - Use, duplication
+//// or disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
+/////////////////////////////////////////////////////////////////////////////////
+
+#include <stdio.h>
+#include <string.h>
+#include <time.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <sys/stat.h>
+
+typedef enum {
+  cp_ebcdic,
+  cp_ascii
+} CodePage;
+
+#define XXXXXX "XXXXXX"
+#define ValidChars "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_.,-"
+typedef struct CodePage {
+  char xes[sizeof(XXXXXX)];
+  char validChars[sizeof(ValidChars)];
+}  CodePage_t;
+
+#pragma convert("IBM-1047")
+static const CodePage_t CharTraitsEbcdic = {
+   XXXXXX,
+   ValidChars
+};
+#pragma convert(pop)
+
+#pragma convert("ISO8859-1")
+static const CodePage_t CharTraitsAscii = {
+   XXXXXX,
+   ValidChars
+};
+#pragma convert(pop)
+
+enum {
+  TempLen = sizeof(XXXXXX)-1
+};
+
+char *real_mkdtemp(CodePage_t cp, char *templ) {
+  if (!templ) {
+    errno=EINVAL;
+    return NULL;
+  }
+  int len = strlen(templ);
+
+  if (len <TempLen || strncmp(templ+len-TempLen,cp.xes ,TempLen)!=0) {
+    errno=EINVAL;
+    return NULL;
+  }
+#define validChars cp.validChars
+  srand(time(NULL));
+#define RANDOM_INT(max) (1 + rand()/((RAND_MAX + 1u) / (max) ))
+#define UPDATE_CHAR(n) templ[len-n+n-1] = validChars[(s##n+c##n)%(sizeof validChars)]
+
+  int s1 = RANDOM_INT(sizeof validChars);
+  int s2 = RANDOM_INT(sizeof validChars);
+  int s3 = RANDOM_INT(sizeof validChars);
+  int s4 = RANDOM_INT(sizeof validChars);
+  int s5 = RANDOM_INT(sizeof validChars);
+  int s6 = RANDOM_INT(sizeof validChars);
+  for (int c1=0; c1<sizeof validChars; ++c1) {
```

Comment:
Can you add a comment to explain what the nested for loops are doing here?

Reply:
Sure.